### ISSUE TITLE
Add `do!` for `load` event

### ISSUE THREAD

**Issue by thedavidmeister** [hoplon/hoplon#90]
I noticed that we don't have a `do!` for `load` which means we can't respond to something (and its children) being fully loaded.

My use case is that I'd like to update a `cell` when my page loads so that I can differentiate between a loaded and not loaded DOM with `cell=`.

**Comment by levitanong** [comment]
I think you can use `with-dom` for that. Can you check it out to see if it does the trick?

**Comment by thedavidmeister** [comment]
@levitanong I can't find any examples of `with-dom` in https://github.com/hoplon/demos and it's completely undocumented in the main repo.

Could you offer an example for me so I can try this approach?

**Comment by levitanong** [comment]
@thedavidmeister 

`with-dom` is a syntactic sugar for `when-dom` (the former takes an element and a body of forms, and the latter takes an element and a function). The example here is in core.cljs itself, in the definition for the default event handling multimethod:

```
(defmethod on! ::default
  [elem event callback]
  (when-dom elem #(.on (js/jQuery elem) (name event) callback)))
```

When-dom polls for the presence of the element in the document, and necessarily means the page has loaded, and the element has been added to the DOM. In the defmethod above, we see that the event listener is bound only when `elem` exists on the DOM.

Another function you might want to look at is `with-init!`, which is a wrapper for `$(function(){dostuff})`. `with-init!` evaluates all the forms in the body once jQuery has loaded.

**Comment by thedavidmeister** [comment]
@levitanong sorry, I don't quite get it still! The details of how to pass in `elem` are confusing me :(

If I have the following DOM:

```
(div
  (img :src "http://images.com/big.jpg"))
```

How do I use `when-dom` to do something when `big.jpg` is loaded?

**Comment by levitanong** [comment]
@thedavidmeister Ah! I completely misunderstood you! I think the mention of "children" made me think of hoplon elements and their children being loaded onto the DOM.

In this case, a `do!` would be the incorrect multimethod to define. We should instead work on implementing an `on!` multimethod. That one is specific to event listeners.

That said, it might not even be necessary to define it. `load` seems to be a generic jQuery event, and the `on!` multi in hoplon has a `::default` that just uses the `(name)` of the keyword that you use, and feeds it to the jQuery `.on` method

In other words, the below should already work.
```clojure
(img :src "http://meatspin"
  :load #(println %))
```

**Comment by thedavidmeister** [comment]
@levitanong from memory, that's the first thing I tried and it didn't work >.< happy to test again though.

**Comment by thedavidmeister** [comment]
I think I just understood the difference between `on!` and `do!` though. Is the idea that if you do `:key value` on a `defelem` attribute and `value` is a function, it runs through `on!` and if it is a value it runs through `do!`?

**Comment by levitanong** [comment]
I do believe that is the case!

**Comment by levitanong** [comment]
Anyway, @thedavidmeister, here is the relevant ::default code.

```clojure
(defmethod on! ::default
  [elem event callback]
  (when-dom elem #(.on (js/jQuery elem) (name event) callback)))
```

This is supposed to catch any event that doesn't have a method that `on!` dispatches on. Here, event is expected to be a keyword, and if you use `:load`, the expression simplifies to:

```clojure
(when-dom elem #(.on (js/jQuery elem) "load" callback))
```

This matches jQuery's `.on( "load", handler )` in https://api.jquery.com/load-event/.

However, if this really doesn't work, I would try:

```clojure
(defmethod on! :load
  [elem _ callback]
  (when-dom elem #(.load (js/jQuery elem) callback)))
```

**Comment by thedavidmeister** [comment]
@levitanong kk, I must have done something wrong. Is `when-dom` used for anything other than `on!`?

**Comment by levitanong** [comment]
@thedavidmeister In the source, it isn't used for anything else. It's used only in the default `on!`.

**Comment by thedavidmeister** [comment]
I think the issue was that it isn't triggered for things that don't contain `img` tags, so the `div` i was trying to check was not working because that doesn't work for jQuery anyway.
